File: tests/fixtures/go/http_handler.go
// Package api provides HTTP handlers for the user service.
package api

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// contextKey is an unexported type for context keys.
type contextKey string

// RequestIDKey is the context key for request IDs.
const RequestIDKey contextKey = "request_id"

// MaxPageSize limits pagination.
var MaxPageSize = 100

// UserService defines the interface for user operations.
type UserService interface {
	GetUser(ctx context.Context, id string) (*User, error)
	ListUsers(ctx context.Context, page int) ([]*User, error)
	CreateUser(ctx context.Context, u *User) error
	DeleteUser(ctx context.Context, id string) error
}

// User represents a user entity.
type User struct {
	ID        string
	Name      string
	Email     string
	CreatedAt time.Time
}

// Handler serves HTTP requests for the user API.
type Handler struct {
	svc    UserService
	logger *Logger
}

// Logger provides structured logging.
type Logger struct {
	prefix string
}

// NewHandler creates a Handler with the given service and logger.
func NewHandler(svc UserService, logger *Logger) *Handler {
	return &Handler{svc: svc, logger: logger}
}

// GetUser handles GET /users/:id.
func (h *Handler) GetUser(w http.ResponseWriter, r *http.Request) {
	id := r.URL.Query().Get("id")
	if id == "" {
		http.Error(w, "missing id", http.StatusBadRequest)
		return
	}

	user, err := h.svc.GetUser(r.Context(), id)
	if err != nil {
		h.logger.logError("GetUser failed", err)
		http.Error(w, "not found", http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(user)
}

// ListUsers handles GET /users.
func (h *Handler) ListUsers(w http.ResponseWriter, r *http.Request) {
	users, err := h.svc.ListUsers(r.Context(), 0)
	if err != nil {
		h.logger.logError("ListUsers failed", err)
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(users)
}

// CreateUser handles POST /users.
func (h *Handler) CreateUser(w http.ResponseWriter, r *http.Request) {
	var user User
	if err := json.NewDecoder(r.Body).Decode(&user); err != nil {
		http.Error(w, "invalid body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	if err := h.svc.CreateUser(r.Context(), &user); err != nil {
		h.logger.logError("CreateUser failed", err)
		http.Error(w, "create failed", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(user)
}

// healthCheck is an unexported handler for internal use.
func (h *Handler) healthCheck(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	fmt.Fprint(w, "ok")
}

// logError logs an error with context.
func (l *Logger) logError(msg string, err error) {
	fmt.Printf("[%s] ERROR: %s: %v\n", l.prefix, msg, err)
}

// logInfo logs an informational message.
func (l *Logger) logInfo(msg string) {
	fmt.Printf("[%s] INFO: %s\n", l.prefix, msg)
}

File: tests/fixtures/python/http_handler.py
"""HTTP handlers for the user API."""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import List, Optional, Dict
import json
import logging

logger = logging.getLogger(__name__)


class UserService(ABC):
    """Interface for user operations."""

    @abstractmethod
    def get_user(self, user_id: str) -> Optional[dict]:
        """Retrieve a user by ID."""

    @abstractmethod
    def list_users(self, page: int = 0) -> List[dict]:
        """List users with pagination."""

    @abstractmethod
    def create_user(self, data: dict) -> dict:
        """Create a new user."""

    @abstractmethod
    def delete_user(self, user_id: str) -> bool:
        """Delete a user by ID."""


@dataclass
class User:
    """User entity."""
    id: str
    name: str
    email: str
    active: bool = True
    metadata: Dict[str, str] = field(default_factory=dict)


class Handler:
    """Serves HTTP requests for the user API."""

    MAX_PAGE_SIZE = 100

    def __init__(self, svc: UserService, debug: bool = False):
        self._svc = svc
        self._debug = debug
        self._request_count = 0

    def get_user(self, request: dict) -> dict:
        """Handle GET /users/:id."""
        user_id = request.get("id")
        if not user_id:
            return {"error": "missing id", "status": 400}

        user = self._svc.get_user(user_id)
        if user is None:
            logger.warning("User not found: %s", user_id)
            return {"error": "not found", "status": 404}

        self._request_count += 1
        return {"data": user, "status": 200}

    def list_users(self, request: dict) -> dict:
        """Handle GET /users."""
        page = request.get("page", 0)
        users = self._svc.list_users(page)
        self._request_count += 1
        return {"data": users, "status": 200}

    def create_user(self, request: dict) -> dict:
        """Handle POST /users."""
        body = request.get("body", {})
        try:
            user = self._svc.create_user(body)
            self._request_count += 1
            return {"data": user, "status": 201}
        except ValueError as e:
            logger.error("Create failed: %s", e)
            return {"error": str(e), "status": 400}

    def _validate_request(self, request: dict) -> bool:
        """Protected: validate incoming request."""
        return "id" in request or "body" in request

    def __reset_counters(self):
        """Private: reset internal counters."""
        self._request_count = 0


class HealthCheck:
    """Simple health check handler."""

    def check(self) -> dict:
        return {"status": "ok"}


def create_app(svc: UserService, debug: bool = False) -> Handler:
    """Factory function to create the application handler."""
    handler = Handler(svc, debug=debug)
    logger.info("Application created, debug=%s", debug)
    return handler


def _setup_logging(level: str = "INFO") -> None:
    """Internal helper for logging setup."""
    logging.basicConfig(level=getattr(logging, level))
